packages/amplication-server/src/core/module/module.service.ts

```diff
@@ -223,6 +228,109 @@ export class ModuleService extends BlockTypeService<
       entityId
     );
 
+    await this.validateDtoReferencesBeforeDeleteModule(moduleId, resourceId);
+    await this.validateActionsDtosReferencesBeforeDeleteModule(
+      moduleId,
+      resourceId
+    );
+
     return super.delete({ where: { id: moduleId } }, user, true); //delete the module and all its children (actions/type...)
   }
+
+  private async validateDtoReferencesBeforeDeleteModule(
+    moduleId: string,
+    resourceId: string
+  ) {
+    const allResourceModuleDtos = await this.moduleDtoService.findMany({
+      where: {
+        resource: {
+          id: resourceId,
+        },
+      },
+    });
+
+    const moduleModuleDtos = allResourceModuleDtos.filter(
+      (moduleDto) => moduleDto.parentBlockId === moduleId
+    );
+
+    const allOtherResourceCustomModuleDtos = allResourceModuleDtos.filter(
+      (moduleDto) =>
+        moduleDto.parentBlockId !== moduleId &&
+        (moduleDto.dtoType === EnumModuleDtoType.Custom ||
+          moduleDto.dtoType === EnumModuleDtoType.CustomEnum)
+    );
+
+    allOtherResourceCustomModuleDtos.forEach((moduleDto) => {
+      moduleDto.properties.forEach((prop) => {
+        if (prop) {
+          prop.propertyTypes.forEach((propType) => {
+            const currentDto = moduleModuleDtos.find(
+              (x) => x.id === propType.dtoId
+            );
+            if (currentDto)
+              throw new AmplicationError(
+                `Cannot delete entity because DTO: ${currentDto.name} is in use in DTO: ${moduleDto.name}.`,
+                { cause: "dtoInUse" }
+              );
+          });
+        }
+      });
+    });
+  }
+
+  private async validateActionsDtosReferencesBeforeDeleteModule(
+    moduleId: string,
+    resourceId: string
+  ) {
+    const moduleModuleDtos = await this.moduleDtoService.findMany({
+      where: {
+        parentBlock: {
+          id: moduleId,
+        },
+        resource: {
+          id: resourceId,
+        },
+      },
+    });
+
+    const allResourceModuleActions = await this.moduleActionService.findMany({
+      where: {
+        resource: {
+          id: resourceId,
+        },
+      },
+    });
+
+    const allOtherResourceCustomModuleActions = allResourceModuleActions.filter(
+      (moduleAction) =>
+        moduleAction.parentBlockId !== moduleId &&
+        moduleAction.actionType === EnumModuleActionType.Custom
+    );
+
+    allOtherResourceCustomModuleActions.forEach((moduleAction) => {
+      if (
+        moduleAction.inputType.type === EnumModuleDtoPropertyType.Dto ||
```

Comment:
Shouldn't we add some more || clauses for the outputType? (perhaps the input is a primitive but the output is a DTO?)

Reply:
@mulygottlieb  Done